src/ansi-c/c_typecheck_code.cpp

```diff
@@ -842,35 +844,147 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
   }
 }
 
+void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+{
+  // compute type
+  typecheck_expr(condition);
+
+  // make it boolean if needed
+  implicit_typecast_bool(condition);
+
+  // non-fatal warnings
+  if(has_subexpr(condition, can_cast_expr<side_effect_expr_function_callt>))
+  {
+    // Remark: we allow function calls without further checks for now because
+    // they are mandatory for some applications.
+    // The next step must be to check that the called functions have a contract
+    // with an empty assigns clause and that they indeed satisfy their contract
+    // using a CI check.
+    warning().source_location = condition.source_location();
+    warning()
+      << "function with possible side-effect called in target's condition"
+      << eom;
+  }
+
+  // fatal errors
+  bool must_throw = false;
+
+  if(condition.type().id() == ID_empty)
+  {
+    must_throw = true;
+    error().source_location = condition.source_location();
+    error() << "void-typed expressions not allowed as assigns clause conditions"
+            << eom;
+  }
```

Comment:
Isn't this caught by `implicit_typecast_bool` already?

Reply:
Since neiter `implicit_typecast_bool`  or `implicity_typecast` have documentation I wasn't sure what to expect.

Reply:
Fair enough! I was actually hoping for your test cases to confirm whether it is/isn't caught...